source/codegen/metadata_validation.py

```diff
@@ -0,0 +1,287 @@
+                if 'type' not in parameter:
+                    if 'grpc_type' not in parameter:
+                        raise Exception(f"parameter {parameter['name']} has no type or grpc_type!")
+                    if not parameter.get('repeated_var_args', False):
+                        raise Exception(f"parameter {parameter['name']} has no type and repeated_var_args is not set!")
+                if 'enum' in parameter:
+                    if parameter['enum'] not in metadata['enums']:
+                        raise Exception(f"parameter {parameter['name']} has enum {parameter['enum']} that was not found!")
+                if 'max_length' in parameter:
+                    if 'repeated_var_args' not in parameter:
+                        raise Exception(f"parameter {parameter['name']} has max_length but no repeated_var_args!")
+                if 'callback_params' in parameter:
+                    for callback_param in parameter['callback_params']:
+                        try:
+                            PARAM_SCHEMA.validate(callback_param)
+                        except Exception as e:
+                            raise Exception(f"Failed to validate callback_param with name {callback_param['name']}")
+                if parameter.get('pointer', False):
+                    # This is technically legal in other cdses but we should only need it for hardcoded values/callback tokens
+                    if 'hardcoded_value' not in parameter and 'callback_token' not in parameter:
+                        raise Exception(f"parameter {parameter['name']} is pointer but not hardcoded_value!")
+                    if parameter.get('include_in_proto', True):
+                        raise Exception(f"parameter {parameter['name']} is pointer but is include_in_proto!")
+
+    except Exception as e:
+        raise Exception(f"Failed to validate function {function_name}") from e
+
+def validate_attribute(attribute: dict, metadata: dict):
+    try:
+        if metadata['config']['module_name'] == 'nisync':
+            # The attributes for nisync only have a name
+            SIMPLE_ATTRIBUTE_SCHEMA.validate(attribute)
+        else:
+            ATTRIBUTE_SCHEMA.validate(attribute)
+        if 'enum' in attribute:
+            if attribute['enum'] not in metadata['enums']:
+                raise Exception(f"attribute {attribute['name']} has enum {attribute['enum']} that was not found!")
+
+    except Exception as e:
+        raise Exception(f"Failed to validate attribute with name {attribute.get('name', '(none)')}") from e
+
+
+def validate_enum(enum_name: str, used_enums: Set[str], metadata: dict):
+    try:
+        enum: Dict[str, Any] = metadata['enums'][enum_name]
+        ENUM_SCHEMA.validate(enum)
+        if enum_name in used_enums:
+            generate_mappings = enum.get('generate-mappings', False)
+            value_types = set([type(value['value']) for value in enum['values']])
+            if not generate_mappings:
```

Comment:
Maybe another case where the conditional could be moved up and have 2 schemas?

Reply:
We could do that here, but that means we'd have to duplicate most of ENUM_SCHEMA, which I think would be less clear/maintainable than doing it this way.